File: apps/ios/ConstructionManager/ConstructionManager/Utilities/FileTypeDetector.swift
//
//  FileTypeDetector.swift
//  ConstructionManager
//
//  Detects file types from file extensions for document viewing
//

import Foundation
import SwiftUI

/// Represents the type of file for viewing purposes
enum ViewableFileType {
    case pdf
    case image
    case video
    case unsupported

    /// System icon name for the file type
    var iconName: String {
        switch self {
        case .pdf: return "doc.fill"
        case .image: return "photo.fill"
        case .video: return "play.rectangle.fill"
        case .unsupported: return "doc.questionmark.fill"
        }
    }

    /// Icon color for the file type
    var iconColor: Color {
        switch self {
        case .pdf: return .red
        case .image: return .blue
        case .video: return .purple
        case .unsupported: return .gray
        }
    }

    /// Human readable label
    var label: String {
        switch self {
        case .pdf: return "PDF Document"
        case .image: return "Image"
        case .video: return "Video"
        case .unsupported: return "Document"
        }
    }
}

/// Utility for detecting file types from file names/extensions
struct FileTypeDetector {

    // Supported image extensions
    private static let imageExtensions: Set<String> = [
        "jpg", "jpeg", "png", "gif", "webp", "bmp", "tiff", "tif", "svg", "heic", "heif"
    ]

    // Supported video extensions
    private static let videoExtensions: Set<String> = [
        "mp4", "mov", "avi", "m4v"
    ]

    /// Detect file type from a file name or path
    /// - Parameter fileName: The file name or full path
    /// - Returns: The detected file type
    static func detect(from fileName: String) -> ViewableFileType {
        let ext = getExtension(from: fileName).lowercased()

        if ext == "pdf" {
            return .pdf
        } else if imageExtensions.contains(ext) {
            return .image
        } else if videoExtensions.contains(ext) {
            return .video
        } else {
            return .unsupported
        }
    }

    /// Extract file extension from a file name or path
    /// - Parameter fileName: The file name or full path
    /// - Returns: The file extension (without dot), or empty string if none
    static func getExtension(from fileName: String) -> String {
        let url = URL(fileURLWithPath: fileName)
        return url.pathExtension
    }

    /// Get MIME type for a file
    /// - Parameter fileName: The file name or full path
    /// - Returns: The MIME type string
    static func getMimeType(from fileName: String) -> String {
        let ext = getExtension(from: fileName).lowercased()

        switch ext {
        case "pdf": return "application/pdf"
        case "jpg", "jpeg": return "image/jpeg"
        case "png": return "image/png"
        case "gif": return "image/gif"
        case "webp": return "image/webp"
        case "bmp": return "image/bmp"
        case "tiff", "tif": return "image/tiff"
        case "svg": return "image/svg+xml"
        case "heic", "heif": return "image/heic"
        case "mp4", "m4v": return "video/mp4"
        case "mov": return "video/quicktime"
        case "avi": return "video/x-msvideo"
        case "doc": return "application/msword"
        case "docx": return "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
        case "xls": return "application/vnd.ms-excel"
        case "xlsx": return "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
        case "ppt": return "application/vnd.ms-powerpoint"
        case "pptx": return "application/vnd.openxmlformats-officedocument.presentationml.presentation"
        case "txt": return "text/plain"
        case "csv": return "text/csv"
        case "zip": return "application/zip"
        default: return "application/octet-stream"
        }
    }

    /// Check if a file can be viewed in-app
    /// - Parameter fileName: The file name or full path
    /// - Returns: True if the file can be viewed in-app
    static func canViewInApp(_ fileName: String) -> Bool {
        let fileType = detect(from: fileName)
        return fileType != .unsupported
    }
}

File: apps/android/construction-android/app/src/main/java/com/constructionpro/app/util/FileTypeDetector.kt
package com.constructionpro.app.util

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import com.constructionpro.app.ui.theme.Primary600
import com.constructionpro.app.ui.theme.ConstructionOrange
import com.constructionpro.app.ui.theme.ConstructionRed

/**
 * File type categories for document viewing
 */
enum class ViewableFileType(
    val label: String,
    val icon: ImageVector,
    val iconColor: Color
) {
    PDF("PDF Document", Icons.Default.PictureAsPdf, ConstructionRed),
    IMAGE("Image", Icons.Default.Image, Primary600),
    VIDEO("Video", Icons.Default.VideoFile, ConstructionOrange),
    UNSUPPORTED("Document", Icons.Default.InsertDriveFile, Color.Gray)
}

/**
 * Utility for detecting file types from file names
 */
object FileTypeDetector {

    private val pdfExtensions = setOf("pdf")

    private val imageExtensions = setOf(
        "jpg", "jpeg", "png", "gif", "webp", "bmp", "tiff", "tif", "svg", "heic", "heif"
    )

    private val videoExtensions = setOf(
        "mp4", "mov", "avi", "mkv", "m4v", "webm", "3gp"
    )

    /**
     * Detect the viewable file type from a filename
     */
    fun detect(fileName: String): ViewableFileType {
        val extension = getExtension(fileName).lowercase()
        return when {
            pdfExtensions.contains(extension) -> ViewableFileType.PDF
            imageExtensions.contains(extension) -> ViewableFileType.IMAGE
            videoExtensions.contains(extension) -> ViewableFileType.VIDEO
            else -> ViewableFileType.UNSUPPORTED
        }
    }

    /**
     * Extract the file extension from a filename
     */
    fun getExtension(fileName: String): String {
        val lastDot = fileName.lastIndexOf('.')
        return if (lastDot >= 0 && lastDot < fileName.length - 1) {
            fileName.substring(lastDot + 1)
        } else {
            ""
        }
    }

    /**
     * Get MIME type for a filename
     */
    fun getMimeType(fileName: String): String {
        val extension = getExtension(fileName).lowercase()
        return when (extension) {
            // PDFs
            "pdf" -> "application/pdf"
            // Images
            "jpg", "jpeg" -> "image/jpeg"
            "png" -> "image/png"
            "gif" -> "image/gif"
            "webp" -> "image/webp"
            "bmp" -> "image/bmp"
            "tiff", "tif" -> "image/tiff"
            "svg" -> "image/svg+xml"
            "heic", "heif" -> "image/heic"
            // Videos
            "mp4" -> "video/mp4"
            "mov" -> "video/quicktime"
            "avi" -> "video/x-msvideo"
            "mkv" -> "video/x-matroska"
            "m4v" -> "video/x-m4v"
            "webm" -> "video/webm"
            "3gp" -> "video/3gpp"
            // Documents
            "doc" -> "application/msword"
            "docx" -> "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
            "xls" -> "application/vnd.ms-excel"
            "xlsx" -> "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
            "ppt" -> "application/vnd.ms-powerpoint"
            "pptx" -> "application/vnd.openxmlformats-officedocument.presentationml.presentation"
            "txt" -> "text/plain"
            "zip" -> "application/zip"
            "rar" -> "application/x-rar-compressed"
            "dwg" -> "application/acad"
            "dxf" -> "application/dxf"
            else -> "application/octet-stream"
        }
    }

    /**
     * Get a human-readable label for the file type
     */
    fun getFileTypeLabel(fileName: String): String {
        val extension = getExtension(fileName).lowercase()
        return when (extension) {
            "pdf" -> "PDF Document"
            "jpg", "jpeg" -> "JPEG Image"
            "png" -> "PNG Image"
            "gif" -> "GIF Image"
            "webp" -> "WebP Image"
            "bmp" -> "Bitmap Image"
            "tiff", "tif" -> "TIFF Image"
            "svg" -> "SVG Image"
            "heic", "heif" -> "HEIC Image"
            "mp4" -> "MP4 Video"
            "mov" -> "QuickTime Video"
            "avi" -> "AVI Video"
            "mkv" -> "MKV Video"
            "m4v" -> "M4V Video"
            "doc", "docx" -> "Microsoft Word Document"
            "xls", "xlsx" -> "Microsoft Excel Spreadsheet"
            "ppt", "pptx" -> "Microsoft PowerPoint"
            "txt" -> "Text File"
            "zip" -> "ZIP Archive"
            "rar" -> "RAR Archive"
            "7z" -> "7-Zip Archive"
            "dwg" -> "AutoCAD Drawing"
            "dxf" -> "DXF Drawing"
            else -> "${extension.uppercase()} File"
        }
    }
}
